File: frameworks/go/gin/Implementation/cache.go
package implementation

import (
	"bytes"
	"net/http"
	"strings"
	"sync"
	"time"

	"github.com/gin-gonic/gin"
)

// cacheRoutes replay a stored answer and skip the handler. Gin ships no response cache, and
// gin-contrib/cache keys on the URL alone, so the vary rows would share one entry. The store and
// the middleware in front of each route are written for it. The handler writes x-rb-serial, so a
// replayed answer repeats the serial it was stored with.
func cacheRoutes(r *gin.Engine, p *Payloads) {
	settings := p.Settings.Cache
	one := headerNames(settings.Vary.One)
	many := headerNames(settings.Vary.Many)
	// rb:wiring cache.*
	store := newStore(settings.Capacity, time.Duration(settings.TTLSeconds)*time.Second)

	r.GET("/cache/small", replay(store, nil), fresh(&p.Small))

	r.GET("/cache/medium", replay(store, nil), fresh(&p.Medium))

	r.GET("/cache/large", replay(store, nil), fresh(&p.Large))

	r.GET("/cache/vary/one", replay(store, one), varied(one), fresh(&p.Small))

	r.GET("/cache/vary/many", replay(store, many), varied(many), fresh(&p.Small))
}

// varied writes the Vary header, which tells a cache in front of the framework what the answer
// depends on. The store keys on the route's own list, not on this header.
func varied(names []string) gin.HandlerFunc {
	vary := strings.Join(names, ", ")
	return func(c *gin.Context) { c.Header("Vary", vary) }
}

// rb:wiring cache.*
// replay answers from the store before the handler runs, and otherwise stores what the handler
// wrote. The key is the path and query, and the request headers the route varies on.
func replay(store *store, vary []string) gin.HandlerFunc {
	return func(c *gin.Context) {
		key := c.Request.URL.RequestURI()
		for _, name := range vary {
			key += "\n" + c.GetHeader(name)
		}
		if hit, ok := store.get(key); ok {
			header := c.Writer.Header()
			for name, values := range hit.header {
				header[name] = values
			}
			c.Status(hit.status)
			_, _ = c.Writer.Write(hit.body)
			c.Abort()
			return
		}
		copied := &copyingWriter{ResponseWriter: c.Writer}
		c.Writer = copied
		c.Next()
		c.Writer = copied.ResponseWriter
		if c.Writer.Status() == http.StatusOK {
			store.put(key, answer{http.StatusOK, c.Writer.Header().Clone(), copied.body.Bytes()})
		}
	}
}

// copyingWriter sends what a handler writes and keeps a copy of it.
type copyingWriter struct {
	gin.ResponseWriter
	body bytes.Buffer
}

func (w *copyingWriter) Write(data []byte) (int, error) {
	w.body.Write(data)
	return w.ResponseWriter.Write(data)
}

func (w *copyingWriter) WriteString(data string) (int, error) {
	w.body.WriteString(data)
	return w.ResponseWriter.WriteString(data)
}

type answer struct {
	status int
	header http.Header
	body   []byte
}

type stored struct {
	answer
	expires time.Time
}

// store is one per process, sized in entries and aged by settings.json. A full store drops one
// entry to take another, whichever the map yields first.
type store struct {
	mu       sync.RWMutex
	entries  map[string]stored
	capacity int
	ttl      time.Duration
}

func newStore(capacity int, ttl time.Duration) *store {
	return &store{entries: make(map[string]stored, capacity), capacity: capacity, ttl: ttl}
}

func (s *store) get(key string) (answer, bool) {
	s.mu.RLock()
	entry, ok := s.entries[key]
	s.mu.RUnlock()
	if !ok || time.Now().After(entry.expires) {
		return answer{}, false
	}
	return entry.answer, true
}

func (s *store) put(key string, a answer) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if _, ok := s.entries[key]; !ok && len(s.entries) >= s.capacity {
		for evicted := range s.entries {
			delete(s.entries, evicted)
			break
		}
	}
	s.entries[key] = stored{a, time.Now().Add(s.ttl)}
}

// rb:end

File: frameworks/node/fastify/Implementation/routes/cache.ts
import type { FastifyReply, FastifyRequest } from "fastify";

import type { Routes } from "../app.ts";
import type { Payload } from "../payloads.ts";
import { answers, payload } from "../schemas.ts";
import { fresh } from "../serial.ts";

declare module "fastify" {
  interface FastifyContextConfig {
    /** The request headers a cache route's answer varies on, which are part of its key. */
    readonly vary?: readonly string[];
  }
}

/** An answer as the handler's reply left it: its headers and the body Fastify serialised. */
interface Stored {
  readonly headers: ReturnType<FastifyReply["getHeaders"]>;
  readonly body: string;
  readonly expires: number;
}

/**
 * cache: the handler skipped and a stored answer written back. The handler writes x-rb-serial,
 * so a replayed answer repeats the serial it was stored with.
 */
const cache: Routes = async (app, { payloads: p }) => {
  const { capacity, ttlSeconds, vary } = p.settings.cache;
  const one = Object.keys(vary.one);
  const many = Object.keys(vary.many);

  // rb:wiring cache.*
  /**
   * Fastify ships no response cache, and @fastify/caching sets cache headers and hands out a
   * store without replaying anything. The replay is two hooks in this family's plugin, so they
   * run on these routes alone: onRequest answers from the store before the handler, and onSend
   * stores what the handler's reply serialised. The store holds settings.json's capacity in
   * entries, and drops the oldest to make room.
   */
  const store = new Map<string, Stored>();

  const keyOf = (request: FastifyRequest): string => {
    let key = request.url;
    for (const header of request.routeOptions.config.vary ?? []) key += `\n${request.headers[header] ?? ""}`;
    return key;
  };

  const lookup = (key: string): Stored | undefined => {
    const hit = store.get(key);
    if (hit !== undefined && hit.expires <= Date.now()) {
      store.delete(key);
      return undefined;
    }
    return hit;
  };

  app.addHook("onRequest", (request, reply, done) => {
    const hit = lookup(keyOf(request));
    if (hit === undefined) done();
    else reply.headers(hit.headers).send(hit.body);
  });

  app.addHook("onSend", (request, reply, body, done) => {
    const key = keyOf(request);
    if (reply.statusCode === 200 && typeof body === "string" && lookup(key) === undefined) {
      if (store.size >= capacity) store.delete(store.keys().next().value!);
      store.set(key, { headers: reply.getHeaders(), body, expires: Date.now() + ttlSeconds * 1000 });
    }
    done(null, body);
  });
  // rb:end

  app.get("/cache/small", answers(payload), async (_request, reply) => stored(reply, p.small));

  app.get("/cache/medium", answers(payload), async (_request, reply) => stored(reply, p.medium));

  app.get("/cache/large", answers(payload), async (_request, reply) => stored(reply, p.large));

  app.get("/cache/vary/one", { ...answers(payload), config: { vary: one } }, async (_request, reply) => stored(reply, p.small, one));

  app.get("/cache/vary/many", { ...answers(payload), config: { vary: many } }, async (_request, reply) => stored(reply, p.small, many));
};

/**
 * The Vary header tells a cache in front of the framework what the answer depends on. The store
 * keys on the route's config, not on this header.
 */
function stored(reply: FastifyReply, answer: Payload, vary?: readonly string[]): Payload {
  if (vary !== undefined) reply.header("vary", vary.join(", "));
  return fresh(reply, answer);
}

export default cache;
